# Homework 03. 

#### MSCA 32013 Optimization and Simulation Methods. University of Chicago.

#### Group 7. Ben Ossyra, Peter Pezon, Yawen Zhang

## Question 1
A refinery produces a high grade (H) and a low grade (L) fuel from five raw stocks with different octane rating. The price and the rating threshold for the two fuel types
are presented in the table below:

|Fuel|Rating|Price ($/B)|
| --- | --- | ---- |
| High |	93	| 37.5 |
| Low |	85	| 28.5 |

The rating, price, and quantity available of each raw material are presented below.

|Stock|Rating|Price|Quantity|
|---|---|----|----|
|  1 |	70	|	9.0	| 2000 |
|  2 |	80	|	12.5	| 4000 |
|  3 |	85	|	12.5	| 4000 |
|  4 |	90	|	27.5	| 5000 |
|  5 |	99	|	27.5 | 3000|


The refinery can sell the high and low fuels and any unused raw materials. The goal is to maximize revenue. Find the optimal blend for the high grade and low grade fuel.



In [2]:
using JuMP, GLPK

m = Model(GLPK.Optimizer)

#Declare Variables:
@variable(m, x1 >= 0)
@variable(m, x2 >= 0)
@variable(m, x3 >= 0)
@variable(m, x4 >= 0)
@variable(m, x5 >= 0)
@variable(m, x6 >= 0)
@variable(m, x7 >= 0)
@variable(m, x8 >= 0)
@variable(m, x9 >= 0)
@variable(m, x10 >= 0)
@variable(m, x11>= 0)
@variable(m, x12>= 0)
@variable(m, x13>= 0)
@variable(m, x14>= 0)
@variable(m, x15>= 0)
@variable(m, x16>= 0)
@variable(m, x17>= 0)

#Add constraints
@constraint(m, constraint_1, x1 + x6 + x11 <= 2000)
@constraint(m, constraint_2, x2 + x7 + x12 <= 4000)
@constraint(m, constraint_3, x3 + x8 + x13 <= 4000)
@constraint(m, constraint_4, x4 + x9 + x14 <= 5000)
@constraint(m, constraint_5, x5 + x10 + x15 <=3000)
@constraint(m, constraint_6, 70*x1+80*x2 +85*x3+90*x4+99*x5>=93*x16)
@constraint(m, constraint_7, 70*x6+80*x7 +85*x8+90*x9+99*x10>=85*x17)
@constraint(m, constraint_8, x1 + x2 + x3 + x4 + x5 == x16)
@constraint(m, constraint_9, x6 + x7 + x8 + x9 + x10 == x17)

#Set the objective

@objective(m, Max, 9*x11+12.5*x12+12.5*x13+27.5*x14+27.5*x15+37.5*x16+28.5*x17)
print(m)


begin
    status = optimize!(m)
end
println("Objective value :", JuMP.objective_value(m))
#println("Optimal solution is x= \n", JuMP.value.(x))
println("x1 = ", JuMP.value(x1))  # => x1 = 1.0
println("x2 = ", JuMP.value(x2))
println("x3 = ", JuMP.value(x3))
println("x4 = ", JuMP.value(x4))
println("x5 = ", JuMP.value(x5))
println("x6 = ", JuMP.value(x6))
println("x7 = ", JuMP.value(x7))
println("x8 = ", JuMP.value(x8))
println("x9 = ", JuMP.value(x9))
println("x10 = ", JuMP.value(x10))
println("x11 = ", JuMP.value(x11))
println("x12 = ", JuMP.value(x12))
println("x13 = ", JuMP.value(x13))
println("x14 = ", JuMP.value(x14))
println("x15 = ", JuMP.value(x15))
println("x16 = ", JuMP.value(x16))
println("x17 = ", JuMP.value(x17))


Objective value :532125.0
x1 = 0.0
x2 = 0.0
x3 = 0.0
x4 = 1416.6666666666702
x5 = 708.3333333333327
x6 = 1999.9999999999998
x7 = 3999.9999999999995
x8 = 4000.0
x9 = 3583.33333333333
x10 = 2291.666666666667
x11 = 0.0
x12 = 0.0
x13 = 0.0
x14 = 0.0
x15 = 0.0
x16 = 2125.0000000000027
x17 = 15874.999999999996


## Question 2

The table below presents hypothetical bid/ask quotes for four major currencies at a given date/time:

| Pair | Bid | Ask |
| --- | --- | ---- |
| EUR/USD |	0.87060	| 0.87063 |
| GBP/USD |	1.42790	| 1.42796 |
| USD/JPY |	133.330	| 133.333 |
| EUR/GBP |	0.60970	 | 0.60972 |
| EUR/JPY |	116.140 |	116.144 |
| GBP/JPY |	190.480 |	190.476 |

The first currency in a currency pair is called the base currency. The bid and the ask are always quoted in terms of the base currency. For example, if you are buying on EUR/USD trade, then you are effectively receiving EUR in exchange for USD at the ask quote.
  
You were given USD  1  to attempt an arbitrage trade, assume no transaction costs.


### A. Write down a linear programming problem to buy and sell quantities of the base currencies and arbitrage/profit of at most USD  10k .  

First you declare all the variables. D is the number of USD. This is what we will try to maximize. To ensure a profit of at most $10,000, we constrain D to 10,001 (the profit plus the dollar we start with). DE is dollars to euros, DG is dollars to pounds, etc.  
<br>
Then we set up our constraints.  
We start with dollars, we set D = 1 (our starting amt) + the currencies exchanged to dollars - the dollars exchanged to other currencies. We want to end up with dollars.  
<br>
Next we set: the euros we are exchanging = euros we have exchanged for. These should cancel out. We don't end up with any euros.
<br>  
We do the same thing we did for euros to pounds and yen.  
<br>
Finally we set the objective to maximize dollars.

### B. Solve for the buy/sell quantities and the arbitrage/profit amount.

In [11]:
# Solve the optimization problem
JuMP.optimize!(m)

# Print the objective value
println("Objective value: ", JuMP.objective_value(m))

println("D = ", JuMP.value.(D))

println("DE = ", JuMP.value.(DE))
println("DG = ", JuMP.value.(DG))
println("DJ = ", JuMP.value.(DJ))

println("ED = ", JuMP.value.(ED))
println("EG = ", JuMP.value.(EG))
println("EJ = ", JuMP.value.(EJ))

println("GD = ", JuMP.value.(GD))
println("GE = ", JuMP.value.(GE))
println("GJ = ", JuMP.value.(GJ))

println("JD = ", JuMP.value.(JD))
println("JE = ", JuMP.value.(JE))
println("JG = ", JuMP.value.(JG))


Objective value: 10001.0
D = 10001.0
DE = 2.0622362181625046e7
DG = 0.0
DJ = 0.0
ED = 0.0
EG = 0.0
EJ = 2.3686712129865784e7
GD = 0.0
GE = 0.0
GJ = 0.0
JD = 2.7509747467626123e9
JE = 0.0
JG = 0.0


### C. Interpret the results and develop a story of the trade executions and the resulting profit.  

You start with 1 USD.  

You exchange the 1 USD for EUR  

In [12]:
EUR = 1 * (1/0.87063) #(USD -> EUR)
print(EUR)

1.148593547201452

Then you exchange your 1.1486 EUR for JPY  

In [13]:
JPY = EUR * 116.140 #(EUR -> JPY)
print(JPY)

133.39765457197663

Then you exchange your 133.3976 JPY back to USD

In [14]:
USD = JPY * (1/133.333) #(JPY -> USD)
print(USD)

1.000484910502101

You have successfully converted 1 USD into ~1.0005 USD. So your abitrage profit is ~0.0005 USD. If you scale this by 20,622,362.18, you will make a profit of $10,000

In [15]:
20622362.181625046 * (USD-1)

10000.0